### DISCUSSION TITLE
Orientation axes

### DISCUSSION THREAD

**Discussion by rodrigomologni** [Kitware/trame#93] [Q&A]
How can I use orientation axes in trame app, please?

**Expectation:**

![image](https://user-images.githubusercontent.com/1608652/178816762-253fde1a-971b-4d80-bc54-8f431e8e5bf8.png)

**Reality:**

![image](https://user-images.githubusercontent.com/1608652/178816549-e2eb1537-d394-4085-a0de-7c3c0bfcb89d.png)

**Source:**

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from trame.app import get_server
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import vtk, vuetify

# -----------------------------------------------------------------------------

from vtkmodules.vtkFiltersSources import vtkConeSource
from vtkmodules.vtkRenderingAnnotation import vtkAxesActor
from vtkmodules.vtkRenderingCore import (
    vtkPolyDataMapper,
    vtkActor,
    vtkRenderer,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
)
from vtkmodules.vtkInteractionWidgets import vtkOrientationMarkerWidget
from vtkmodules.vtkInteractionStyle import vtkInteractorStyleSwitch  # noqa
from vtkmodules import vtkRenderingOpenGL2  # noqa

source = vtkConeSource()

mapper = vtkPolyDataMapper()
mapper.SetInputConnection(source.GetOutputPort())

actor = vtkActor()
actor.SetMapper(mapper)
actor.GetProperty().EdgeVisibilityOn()

renderer = vtkRenderer()
renderer.AddActor(actor)
renderer.SetBackground(.3, .3, .3)
renderer.ResetCamera()

window = vtkRenderWindow()
window.AddRenderer(renderer)
window.SetSize(800, 600)

interactor = vtkRenderWindowInteractor()
interactor.SetRenderWindow(window)
interactor.GetInteractorStyle().SetCurrentStyleToTrackballCamera()

widget = vtkOrientationMarkerWidget()
widget.SetOrientationMarker(vtkAxesActor())
widget.SetInteractor(interactor)
widget.EnabledOn()
widget.InteractiveOff()

# interactor.Start()

# -----------------------------------------------------------------------------

server = get_server()
ctrl = server.controller

with SinglePageLayout(server) as layout:
    with layout.content:
        with vuetify.VContainer(fluid=True, classes="pa-0 fill-height"):
            html_view = vtk.VtkLocalView(window)
            ctrl.on_server_ready.add(html_view.update)

if __name__ == "__main__":
    server.start()

```

**Output:**
```
And to prevent your browser from opening, add '--server' to your command line.
      !!!No serializer for vtkActor2D with id 0000027968121ce0
      !!!No serializer for vtkAxesActor with id 0000027968029c30

App running at:
 - Local:   http://localhost:8080/
 - Network: http://172.22.224.1:8080/

Note that for multi-users you need to use and configure a launcher.
```

**Comment by jourdain** [comment]
Like the error says, `vtkActor2D` and `vtkAxesActor` serializers needs to be registered to properly work with`VtkLocalView`.

**Reply by jourdain** [reply]
Those are not yet implemented in the mainstream package but should since they are mostly available on the vtk.js side.

**Reply by rodrigomologni** [reply]
Hi @jourdain!

I didn't try to run the code remotely because I thought `VtkLocalView` and `VtkRemoteView` would produce the same result.

Is it possible that this feature will be implemented soon? I would like to run my app locally too.

Thank you for answer me.

**Reply by jourdain** [reply]
What do you mean by running your app locally? Both implementation can work locally and remotely. VtkRemoteView just mean that the UI gets an ImageStream from VTK rather than some Geometry.

**Reply by rodrigomologni** [reply]
I would like to change the view mode (local/remote), like in the [MultiFilter](https://github.com/Kitware/trame/tree/master/examples/06_vtk/Applications/MultiFilter) example. However, the orientation axes are not shown in the local mode, only in the remote mode.

**Reply by rodrigomologni** [reply]
Hi @jourdain! I decided to use only the remote view in my app because it meets all my needs. Thank you very much for your attention. 🙏

**Reply by pkitslaar** [reply]
> Those are not yet implemented in the mainstream package but should since they are mostly available on the vtk.js side.

What would be needed to add this. Are there any guides how this can be added?

**Reply by jourdain** [reply]
Not much, it would require some update mainly in [VTK](https://github.com/Kitware/VTK/blob/master/Web/Python/vtkmodules/web/render_window_serializer.py) which can be overridden within [trame itself over here](https://github.com/Kitware/trame-vtk/blob/master/trame_vtk/modules/vtk/addon_serializer.py) to test/deploy it sooner. Ideally the migration into VTK proper will be done once the code has been validated and tested.
